src/core/row.cpp

```diff
@@ -16,6 +16,26 @@
 using namespace soci;
 using namespace details;
 
+namespace soci
+{
+    namespace details
+    {
+        struct data_holder
+        {
```

Comment:
Why not give this struct a ctor (taking `type` value and initializing `ind` to `i_ok`) and dtor (cleaning up `data.p` for strings and times)? It looks like it could simplify the code quite a bit.